### PR TITLE
Implemented Sagemaker Batch Transform integration with MLflow

### PR DESCRIPTION
**Pull request by YQ-Wang**
## What changes are proposed in this pull request?
Implemented the Python APIs and the CLI supports to integrate Sagemaker Batch Transform with MLflow.
https://docs.aws.amazon.com/sagemaker/latest/dg/batch-transform.html

`deploy_transform_job()` is used to deploy the model from the MLflow model registry as a Sagemaker Batch Transform job.
`terminate_transform_job()` is used to terminate the in-progress Sagemaker Batch Transform job.

## How is this patch tested?
- Unit tests
- Checked linter by running lint.sh
- Manually used the API to deploy a model as a Sagemaker Batch Transform job and checked it from the AWS Console
- Manually used the API to terminate the deployed Sagemaker Batch Transform job and checked it from the AWS Console
- Generated the docs and checked it.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

The model can be deployed as a Sagemaker Batch Transform job using Pythons APIs and CLIs.

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [x] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by YQ-Wang** [comment]
Hey @dbczumar, I accidentally closed the old PR [#4304](https://github.com/mlflow/mlflow/pull/4304) while fixing some comments length issues. Here is the updated code.  Thank you!

**Comment by YQ-Wang** [comment]
Hey @dbczumar and @tomasatdatabricks , please let me know if I need to do anything at this point. Thank you!

**Comment by petervandenabeele** [comment]
Hi, is there any indication of the timeframe for this functionality to be merged? Thanks.

**Comment by YQ-Wang** [comment]
Hi @petervandenabeele, the unit test for this PR has been conducted. I am also actively waiting for feedback from them.

**Comment by petervandenabeele** [comment]
Hi, I was asked to ping @dbczumar and @tomasatdatabricks for this PR. My customer is using MLFlow actively and we now also want to use it in combination with SageMaker batch inference. So, it would be nice to see this merged.

Also, @YQ-Wang is it a problem for merging that 4 checks are failing in the list above (not clear to me how critical that is).

**Comment by YQ-Wang** [comment]
Hi @petervandenabeele, I checked the logs earlier. It seems to me the 4 failed test cases are not relevant to this PR.

**Review by dbczumar — COMMENTED** [review]
@YQ-Wang This is fantastic! Huge apologies for the delay here. I tried this out manually using the MLflow model trained via https://github.com/mlflow/mlflow/blob/master/examples/sklearn_elasticnet_wine/train.py with a JSON representation of the wine quality dataset:

```
 mlflow sagemaker deploy-transform-job -n mytestjob10 -m mlruns/0/c3e11313721f4ee58c88f3a646b3650a/artifacts/model/ -dt S3Prefix -in s3://<bucket_name_redacted>/wine-quality.json --content-type application/json -out s3://<bucket_name_redacted>/sagemakerresults --image-url <image_url_redacted> -b <bucket_name_redacted> -e <role_name_redacted>
````

This successfully created a SageMaker batch transform inference job:

Screen Shot 2021-07-19 at 1 54 56 AM

I also launched a second job (`myjob11`) and terminated synchronously via `mlflow sagemaker terminate-transform-job -n myjob11`. This termination succeeded:

```
$ mlflow sagemaker terminate-transform-job -n myjob11
2021/07/19 01:32:27 INFO mlflow.sagemaker: Terminated batch transform job with <arn_redacted>
2021/07/19 01:32:27 INFO mlflow.sagemaker: Waiting for the termination operation to complete...
2021/07/19 01:32:27 INFO mlflow.sagemaker: Termination is still in progress. Current batch transform job status:                    Stopping
2021/07/19 01:32:47 INFO mlflow.sagemaker: Termination is still in progress. Current batch transform job status:                    Stopping
2021/07/19 01:33:08 INFO mlflow.sagemaker: Termination is still in progress. Current batch transform job status:                    Stopping
2021/07/19 01:33:28 INFO mlflow.sagemaker: Termination is still in progress. Current batch transform job status:                    Stopping
2021/07/19 01:33:48 INFO mlflow.sagemaker: Termination is still in progress. Current batch transform job status:                    Stopping
2021/07/19 01:34:08 INFO mlflow.sagemaker: Termination is still in progress. Current batch transform job status:                    Stopping
2021/07/19 01:34:29 INFO mlflow.sagemaker: Termination is still in progress. Current batch transform job status:                    Stopping
2021/07/19 01:34:49 INFO mlflow.sagemaker: Termination is still in progress. Current batch transform job status:                    Stopping
2021/07/19 01:34:54 INFO mlflow.sagemaker: The termination operation completed successfully with message: "The SageMaker batch transform job was terminated successfully."
```

I did run into the following model inference issue once the batch transform job ran:

```
{"error_code": "BAD_REQUEST", "message": "Encountered an unexpected error while evaluating the model. Verify that the serialized input Dataframe is compatible with the model for inference.", "stack_trace": "Traceback (most recent call last):\n  File \"/opt/mlflow/mlflow/pyfunc/scoring_server/__init__.py\", line 264, in transformation\n    raw_predictions = model.predict(data)\n  File \"/opt/mlflow/mlflow/pyfunc/__init__.py\", line 596, in predict\n    return self._model_impl.predict(data)\n  File \"/miniconda/envs/custom_env/lib/python3.8/site-packages/sklearn/linear_model/_base.py\", line 238, in predict\n    return self._decision_function(X)\n  File \"/miniconda/envs/custom_env/lib/python3.8/site-packages/sklearn/linear_model/_coordinate_descent.py\", line 898, in _decision_function\n    return super()._decision_function(X)\n  File \"/miniconda/envs/custom_env/lib/python3.8/site-packages/sklearn/linear_model/_base.py\", line 221, in _decision_function\n    return safe_sparse_dot(X, self.coef_.T,\n  File \"/
```

Here's what the content of `wine-quality.json` looked like:

```
{"columns":["fixed acidity","volatile acidity","citric acid","residual sugar","chlorides","free sulfur dioxide","total sulfur dioxide","density","pH","sulphates","alcohol","quality"],"index":[0,1,2,3,4],"data":[[7.0,0.27,0.36,20.7,0.045,45.0,170.0,1.001,3.0,0.45,8.8,6],[6.3,0.3,0.34,1.6,0.049,14.0,132.0,0.994,3.3,0.49,9.5,6],[8.1,0.28,0.4,6.9,0.05,30.0,97.0,0.9951,3.26,0.44,10.1,6],[7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.4,9.9,6],[7.2,0.23,0.32,8.5,0.058,47.0,186.0,0.9956,3.19,0.4,9.9,6]]}
```

This matches the format corresponding to the `application/json` MIME type described here: https://mlflow.org/docs/latest/models.html#deploy-mlflow-models. I also tried this with a CSV-formatted input dataset and had the same issue.

Are there any extra steps that users need to take in order to successfully perform inference? If not, for PR review purposes, do you have an example model + dataset that works end-to-end? In the future, it would be awesome to include this as an example in the [examples](https://github.com/mlflow/mlflow/tree/master/examples) directory, but we can follow up with that later.

Once I can confirm that inference works end-to-end and a couple of small formatting / nit comments have been addressed, I'd be very excited to merge this. Thanks a bunch for your contribution!

**Comment by YQ-Wang** [comment]
Hey @dbczumar, thank you for taking the time to review my PR! I have committed the suggested modifications. For the model inference issue you mentioned above, I think it is probably due to the redundant column names. The acceptable data format is shown [here](https://docs.aws.amazon.com/sagemaker/latest/dg/cdf-inference.html). We need to remove the headers before we use the data for batch inference. Another thing I can think of is for the image URL - you need to use the build-and-push-container cli to generate an image in ECR. Please give me some time to provide an example for you. Thanks again.

File: mlflow/sagemaker/__init__.py
Comment (dbczumar) [reply mlflow/sagemaker/__init__.py:494]: Reopening here: Can we add the `experimental` tag to the new public APIs in this file as well?

**Review by dbczumar — APPROVED** [review]
@YQ-Wang Thanks for addressing the feedback so quickly! LGTM once https://github.com/mlflow/mlflow/pull/4410/files#r672106177 has been addressed (just a couple of missing `@experimental` tags). I confirmed that dropping the index and header on the input data used for testing resulted in a successful batch transform execution.

Before merging, can you also sign your commits in order to pass our DCO check? https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.rst#sign-your-work

File: mlflow/sagemaker/__init__.py
Comment (YQ-Wang) [reply mlflow/sagemaker/__init__.py:496]: Sure thing. I just added the experimental tag.

**Comment by dbczumar** [comment]
Merging! Thanks again @YQ-Wang !

**Pull request merged**
